lib/puppet/util/ssl.rb

```diff
@@ -65,26 +65,22 @@ def self.is_possibly_valid_dn?(dn)
   #   invalidated the connection
   # @param [String] host The DNS name of the other end of the SSL connection
   #
+  # @deprecated Use `Puppet::SSL::Verifier` instead
   # @raises [Puppet::Error, OpenSSL::OpenSSLError]
   def self.handle_connection_error(error, verifier, host)
     # can be nil
     peer_cert = verifier.peer_certs.last
 
-    if error.message.include? "certificate verify failed"
+    # If cert/hostname are mismatched, openssl 1.1+ will generate a
+    # `verify_error` whose error code is `V_OK`, since nothing is wrong with the
+    # cert chain itself. So check for mismatch ourselves using the same routine
+    # that ruby's `post_connection_check` would do (and does with openssl 1.0)
```

Comment:
Turns out the issue is not openssl 1.1, but rather ruby/openssl introduced a new `:verify_hostname` option, which defaults to true. When enabled, ruby will automatically perform hostname verification. This is generally a good thing, since applications don't need to remember to call `post_connection_check`, but in our case, it breaks our cert-mismatch detection. Need to rework the test and update these comments.